torch/distributed/fsdp/shard_utils.py

```diff
@@ -0,0 +1,144 @@
+
+    return offsets
+
+
+def _offsets_to_split_sizes(
+    input_offsets: List[int],
+    output_offsets: List[int],
+    tensor_numel: int,
+    world_size: int,
+    my_rank: int,
+) -> Tuple[List[int], List[int]]:
+    r"""
+    Given the shard offsets for each rank of the input tensor and output tensor,
+    this API returns the corresponding split sizes that can be passed to
+    all_to_all_single().
+    """
+
+    def _get_interval(offsets):
+        if my_rank != world_size - 1:
+            return offsets[my_rank], offsets[my_rank + 1] - 1
+        else:
+            return offsets[my_rank], tensor_numel - 1
+
+    def _offsets_to_sizes(offsets, begin, end):
+        sizes = []
+        for i, offset in enumerate(offsets):
+            next_offset = offsets[i + 1] if i < len(offsets) - 1 else end + 1
+            sizes.append(
+                (next_offset - offset)
+                - max(begin - offset, 0)
+                - max(next_offset - end - 1, 0)
+            )
+        return sizes
+
+    def _convert(from_offsets, to_offsets, split_sizes):
+        begin, end = _get_interval(from_offsets)
+        to_begin_rank = bisect.bisect(to_offsets, begin) - 1
+        to_end_rank = bisect.bisect(to_offsets, end) - 1
+        _split_sizes = _offsets_to_sizes(
+            to_offsets[to_begin_rank : to_end_rank + 1], begin, end
+        )
+        split_sizes[to_begin_rank : to_end_rank + 1] = _split_sizes
+
+    input_split_sizes = [0 for _ in range(world_size)]
+    output_split_sizes = [0 for _ in range(world_size)]
+    _convert(input_offsets, output_offsets, input_split_sizes)
+    _convert(output_offsets, input_offsets, output_split_sizes)
+
+    return input_split_sizes, output_split_sizes
+
+
+def reshard_flatten_tensor(
+    input_tensor: ShardedTensor,
+    output_spec: ShardingSpec,
+    world_size: int,
+    my_rank: int,
+    device: torch.device,
+    process_group: Optional[ProcessGroup],
+) -> torch.Tensor:
+    """
+    Resharded a sharded flatten tensor, this is used by FSDP to do sharded
+    state_dict. But the functionaility is not supported by ShardedTensor.
+    This API is designed to be used for FSDP; therefore this API supports only
+    1-D ShardedTensor (hence the naming, reshard_flatten_tensor).
+
+    This API uses the ChunkShardingSpec and EnumerableShardingSpec from
+    torch.distributed.sharding_spec but ignores the placement field in
+    ChunkShardingSpec, as the placement requires the callees understand the
+    number of GPUs per node. The API simply uses the semantics of the sharding
+    specs.
+
+    Args:
+        input_tensor (ShardedTensor): the original ShardedTensor. Must be 1D.
+        output_spec (ShardingSpec): the sharding spect for the output tensor.
+        world_size (int): total trainer count.
+        my_rank (int): the rank for this trainer.
+
+    Returns:
+        The local shard for the new ShardedTensor.
```

Comment:
Is there a reason we're returning a torch.Tensor which is the local shard instead of a ShardedTensor instance?